### PR TITLE
[SPARK-15073][SQL] Hide SparkSession constructor from the public

### PR DESCRIPTION
**Pull request by andrewor14**
## What changes were proposed in this pull request?

Users should use the builder pattern instead.
## How was this patch tested?

Jenks.

### CODE REVIEW

**Comment by andrewor14** [comment]
retest this please

**Comment by rxin** [comment]
LGTM pending Jenkins.

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 1]
```diff
@@ -63,7 +64,7 @@ class SparkSession private(
     @transient private val existingSharedState: Option[SharedState])
   extends Serializable with Logging { self =>
 
-  def this(sc: SparkContext) {
+  private[sql] def this(sc: SparkContext) {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: actually is it possible to just make this private?

**Comment by rxin** [comment]
Merging in master/branch-2.0.

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 1 above)
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: we use this in SQLContext

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: maybe we should change that and have sqlcontext always getting it from sparksession tiself?

Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: yeah I think the problem is that in `getOrCreate` right now we go the other way, i.e. getting the session from the context

File: repl/scala-2.11/src/main/scala/org/apache/spark/repl/Main.scala [hunk 2]
```diff
@@ -94,10 +94,10 @@ object Main extends Logging {
 
   def createSparkSession(): SparkSession = {
     if (SparkSession.hiveClassesArePresent) {
-      sparkSession = SparkSession.withHiveSupport(sparkContext)
+      sparkSession = SparkSession.builder.enableHiveSupport().getOrCreate()
```
Comment (cloud-fan) [inline repl/scala-2.11/src/main/scala/org/apache/spark/repl/Main.scala]: They are not identical. The previous one will create a new spark session anyway, but the new one may not create a new spark session if sparkContext already exists. Then we can't guarantee the returned spark session supports hive.

Comment (andrewor14) [inline repl/scala-2.11/src/main/scala/org/apache/spark/repl/Main.scala]: good catch

**Pull request closed without merging**